pydm/data_plugins/local_plugin.py

```diff
@@ -1,332 +1,427 @@
-            and slots.
-
-            :param channel: The channel to connect.
-            :type channel:  :class:`PyDMChannel`
-            """
-            super(Connection, self).add_listener(channel)
-            if self.is_connected():
-                self.send_connection_state(conn=True)
-                self.update()
+                self.prec_signal.emit(self._precision_set)
+
+        if channel.connection_slot is not None:
+            self.send_connection_state(conn=True)
+
+        # Connect the put_value slot to the channel's value_signal,
+        # which captures the values sent through the plugin
+        if channel.value_signal is not None:
             try:
-                channel.value_signal[str].connect(self.put_value, Qt.QueuedConnection)
-                channel.value_signal[int].connect(self.put_value, Qt.QueuedConnection)
-                channel.value_signal[float].connect(self.put_value, Qt.QueuedConnection)
-            except:
+                channel.value_signal[int].connect(
+                    self.put_value, Qt.QueuedConnection)
+            except KeyError:
                 pass
             try:
-                channel.waveform_signal.connect(self.put_value, Qt.QueuedConnection)
-            except:
+                channel.value_signal[float].connect(
+                    self.put_value, Qt.QueuedConnection)
+            except KeyError:
                 pass
+            try:
+                channel.value_signal[str].connect(
+                    self.put_value, Qt.QueuedConnection)
+            except KeyError:
+                pass
+            try:
+                channel.value_signal[bool].connect(
+                    self.put_value, Qt.QueuedConnection)
+            except KeyError:
+                pass
+            try:
+                channel.value_signal[np.ndarray].connect(
+                    self.put_value, Qt.QueuedConnection)
+            except KeyError:
+                pass
+
+    @Slot(int)
+    @Slot(float)
+    @Slot(str)
+    @Slot(bool)
+    @Slot(np.ndarray)
+    def put_value(self, new_value):
+        """
+        Slot connected to the channal.value_signal.
+        Updates the value of this local variable and then broadcasts it to
+        the other listeners to this channel
+        """
+        if new_value is not None:
+            # update the attributes here with the new values
+            self.value = new_value
+            # send this value
+            self.send_new_value(new_value)
+            # send precision for float values
+            if isinstance(new_value, float):
+                if self._precision_set is None:
+                    self.send_precision(new_value)
+                else:
+                    self.prec_signal.emit(self._precision_set)
+
+
+class LocalPlugin(PyDMPlugin):
+    protocol = "loc"
+    connection_class = Connection
+
+    @staticmethod
+    def get_connection_id(channel):
```

Comment:
I can't help but notice that this method looks very similar to code you've written earlier in the file. You should consider if there's an easy way to write a function that handles the shared functionality.